In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import imageio

import seaborn as sns

## Loading and inspecting the fertility_rates

In [2]:
fert = pd.read_csv('./data/gapminder_total_fertility.csv', index_col=0)

In [3]:
fert.shape

(260, 216)

In [4]:
fert.head()

1800  1801  1802  1803  1804  1805  1806  1807  1808  \
Total fertility rate                                                          
Abkhazia                NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Afghanistan            7.00  7.00  7.00  7.00  7.00  7.00  7.00  7.00  7.00   
Akrotiri and Dhekelia   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Albania                4.60  4.60  4.60  4.60  4.60  4.60  4.60  4.60  4.60   
Algeria                6.99  6.99  6.99  6.99  6.99  6.99  6.99  6.99  6.99   

                       1809  ...  2006  2007  2008  2009  2010  2011  2012  \
Total fertility rate         ...                                             
Abkhazia                NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Afghanistan            7.00  ...  6.70  6.46  6.20  5.93  5.66  5.40  5.14   
Akrotiri and Dhekelia   NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Albania                4.60  ...  1.85  1.80  1.76  1.74  1.74  1.75  1.76   
Algeria                6.99  ...  2.58  2.66  2.73  2.78  2.82  2.83  2.82   

                       2013  2014  2015  
Total fertility rate                     
Abkhazia                NaN   NaN   NaN  
Afghanistan            4.90  4.68  4.47  
Akrotiri and Dhekelia   NaN   NaN   NaN  
Albania                1.77  1.78  1.78  
Algeria                2.80  2.76  2.71  

[5 rows x 216 columns]

In [5]:
fert.tail()

1800  1801  1802  1803  1804  1805  1806  1807  1808  \
Total fertility rate                                                         
Yugoslavia             NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Zambia                6.71  6.71  6.71  6.71  6.71  6.71  6.71  6.71  6.71   
Zimbabwe              6.75  6.75  6.75  6.75  6.75  6.75  6.75  6.75  6.75   
Åland                  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Åland                  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

                      1809  ...  2006  2007  2008  2009  2010  2011  2012  \
Total fertility rate        ...                                             
Yugoslavia             NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Zambia                6.71  ...  5.93  5.91  5.88  5.85  5.81  5.77  5.73   
Zimbabwe              6.75  ...  3.94  3.90  3.85  3.79  3.72  3.64  3.56   
Åland                  NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Åland                  NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

                      2013  2014  2015  
Total fertility rate                    
Yugoslavia             NaN   NaN   NaN  
Zambia                5.69  5.64  5.59  
Zimbabwe              3.49  3.41  3.35  
Åland                  NaN   NaN   NaN  
Åland                  NaN   NaN   NaN  

[5 rows x 216 columns]

**Observation 1: The last two rows are indexed "Åland". This is probably a mistake.**

In [6]:
#Changing column names of life to integers

fert.columns = fert.columns.astype(int)

## Loading and inspecting the life_expectancy

In [7]:
life = pd.read_excel('./data/gapminder_lifeexpectancy.xlsx', index_col=0)

In [8]:
life.shape

(260, 217)

**Observation 2: The life-dataframe has one more column ("2016") compared to the fert-dataframe**

In [9]:
life.head()

1800.0  1801.0  1802.0  1803.0  1804.0  1805.0  1806.0  \
Life expectancy                                                                 
Abkhazia                  NaN     NaN     NaN     NaN     NaN     NaN     NaN   
Afghanistan             28.21   28.20   28.19   28.18   28.17   28.16   28.15   
Akrotiri and Dhekelia     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
Albania                 35.40   35.40   35.40   35.40   35.40   35.40   35.40   
Algeria                 28.82   28.82   28.82   28.82   28.82   28.82   28.82   

                       1807.0  1808.0  1809.0  ...  2007.0  2008.0  2009.0  \
Life expectancy                                ...                           
Abkhazia                  NaN     NaN     NaN  ...     NaN     NaN     NaN   
Afghanistan             28.14   28.13   28.12  ...    52.4    52.8    53.3   
Akrotiri and Dhekelia     NaN     NaN     NaN  ...     NaN     NaN     NaN   
Albania                 35.40   35.40   35.40  ...    76.6    76.8    77.0   
Algeria                 28.82   28.82   28.82  ...    75.3    75.5    75.7   

                       2010.0  2011.0  2012.0  2013.0  2014.0  2015.0  2016.0  
Life expectancy                                                                
Abkhazia                  NaN     NaN     NaN     NaN     NaN     NaN     NaN  
Afghanistan              53.6    54.0    54.4    54.8    54.9    53.8   52.72  
Akrotiri and Dhekelia     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
Albania                  77.2    77.4    77.5    77.7    77.9    78.0   78.10  
Algeria                  76.0    76.1    76.2    76.3    76.3    76.4   76.50  

[5 rows x 217 columns]

In [10]:
life.tail()

1800.0  1801.0  1802.0  1803.0  1804.0  1805.0  1806.0  \
Life expectancy                                                           
Yugoslavia          NaN     NaN     NaN     NaN     NaN     NaN     NaN   
Zambia            32.60   32.60   32.60   32.60   32.60   32.60   32.60   
Zimbabwe          33.70   33.70   33.70   33.70   33.70   33.70   33.70   
Åland               NaN     NaN     NaN     NaN     NaN     NaN     NaN   
South Sudan       26.67   26.67   26.67   26.67   26.67   26.67   26.67   

                 1807.0  1808.0  1809.0  ...  2007.0  2008.0  2009.0  2010.0  \
Life expectancy                          ...                                   
Yugoslavia          NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN   
Zambia            32.60   32.60   32.60  ...    49.0    51.1    52.3    53.1   
Zimbabwe          33.70   33.70   33.70  ...    46.4    47.3    48.0    49.1   
Åland               NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN   
South Sudan       26.67   26.67   26.67  ...    55.5    55.6    55.8    56.0   

                 2011.0  2012.0  2013.0  2014.0  2015.0  2016.0  
Life expectancy                                                  
Yugoslavia          NaN     NaN     NaN     NaN     NaN     NaN  
Zambia             53.7    54.7    55.6    56.3    56.7   57.10  
Zimbabwe           51.6    54.2    55.7    57.0    59.3   61.69  
Åland               NaN     NaN     NaN     NaN     NaN     NaN  
South Sudan        55.9    56.0    56.0    56.1    56.1   56.10  

[5 rows x 217 columns]

**Observation 3: The last row of the life-dataframe is indexed with "South Sudan".**

In [11]:
#Changing column namnes of life to integers

life.columns = life.columns.astype(int)

## Tidy the fert and life dataframes

In [12]:
#Adjustment 1: Check whether the index values show further deviations

In [13]:
# Create a list with the index vales of fert and life

fert_index_list = fert.index.to_list()
life_index_list = life.index.to_list()

In [14]:
difference_index = [index for index in life_index_list if index not in fert_index_list]
difference_index

['South Sudan']

**Observation 4: "South Sudan" is indeed the only deviation between the indexes of the fert and life dataframes**

In [15]:
fert.iloc[258:259].value_counts()
fert.iloc[259:].value_counts()

Series([], dtype: int64)

Way to solve the issue: The last two lines don't include any values. Since they will not have any impact on our analysis, I will change the index of the last row to "South Sudan" like in the life-dataframe. I assume that this will make it easier to merge them.

In [16]:
fert_index_list = fert.index.to_list()
fert_index_list[-1] = "South Sudan"
fert.index = fert_index_list

In [17]:
fert.tail()

1800  1801  1802  1803  1804  1805  1806  1807  1808  1809  ...  \
Yugoslavia    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
Zambia       6.71  6.71  6.71  6.71  6.71  6.71  6.71  6.71  6.71  6.71  ...   
Zimbabwe     6.75  6.75  6.75  6.75  6.75  6.75  6.75  6.75  6.75  6.75  ...   
Åland         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
South Sudan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   

             2006  2007  2008  2009  2010  2011  2012  2013  2014  2015  
Yugoslavia    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
Zambia       5.93  5.91  5.88  5.85  5.81  5.77  5.73  5.69  5.64  5.59  
Zimbabwe     3.94  3.90  3.85  3.79  3.72  3.64  3.56  3.49  3.41  3.35  
Åland         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
South Sudan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 216 columns]

In [18]:
#Adjustment 2: Add an additional column

fert[2016] = fert[2015]

Way to solve the issue: Copy the values from 2015 and create a new column "2016" in the fert dataframe. Another approach would be to drop the "2016" column in the life-dataframe.

In [19]:
fert.shape

(260, 217)

In [20]:
fert.shape

(260, 217)

## Step 5

In [21]:
fert.index

Index(['Abkhazia', 'Afghanistan', 'Akrotiri and Dhekelia', 'Albania',
       'Algeria', 'American Samoa', 'Andorra', 'Angola', 'Anguilla',
       'Antigua and Barbuda',
       ...
       'Vietnam', 'Virgin Islands (U.S.)', 'North Yemen (former)',
       'South Yemen (former)', 'Yemen', 'Yugoslavia', 'Zambia', 'Zimbabwe',
       'Åland', 'South Sudan'],
      dtype='object', length=260)

In [22]:
fert.index.name = 'country'

## Step 6

In [23]:
fert = fert.reset_index()

In [24]:
fert = fert.melt(id_vars='country', var_name='year', value_name='fertility_rate')

In [25]:
fert

country  year  fertility_rate
0                   Abkhazia  1800             NaN
1                Afghanistan  1800            7.00
2      Akrotiri and Dhekelia  1800             NaN
3                    Albania  1800            4.60
4                    Algeria  1800            6.99
...                      ...   ...             ...
56415             Yugoslavia  2016             NaN
56416                 Zambia  2016            5.59
56417               Zimbabwe  2016            3.35
56418                  Åland  2016             NaN
56419            South Sudan  2016             NaN

[56420 rows x 3 columns]

## Step 7

In [26]:
life

1800   1801   1802   1803   1804   1805   1806   1807  \
Life expectancy                                                                 
Abkhazia                 NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
Afghanistan            28.21  28.20  28.19  28.18  28.17  28.16  28.15  28.14   
Akrotiri and Dhekelia    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
Albania                35.40  35.40  35.40  35.40  35.40  35.40  35.40  35.40   
Algeria                28.82  28.82  28.82  28.82  28.82  28.82  28.82  28.82   
...                      ...    ...    ...    ...    ...    ...    ...    ...   
Yugoslavia               NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
Zambia                 32.60  32.60  32.60  32.60  32.60  32.60  32.60  32.60   
Zimbabwe               33.70  33.70  33.70  33.70  33.70  33.70  33.70  33.70   
Åland                    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
South Sudan            26.67  26.67  26.67  26.67  26.67  26.67  26.67  26.67   

                        1808   1809  ...  2007  2008  2009  2010  2011  2012  \
Life expectancy                      ...                                       
Abkhazia                 NaN    NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
Afghanistan            28.13  28.12  ...  52.4  52.8  53.3  53.6  54.0  54.4   
Akrotiri and Dhekelia    NaN    NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
Albania                35.40  35.40  ...  76.6  76.8  77.0  77.2  77.4  77.5   
Algeria                28.82  28.82  ...  75.3  75.5  75.7  76.0  76.1  76.2   
...                      ...    ...  ...   ...   ...   ...   ...   ...   ...   
Yugoslavia               NaN    NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
Zambia                 32.60  32.60  ...  49.0  51.1  52.3  53.1  53.7  54.7   
Zimbabwe               33.70  33.70  ...  46.4  47.3  48.0  49.1  51.6  54.2   
Åland                    NaN    NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   
South Sudan            26.67  26.67  ...  55.5  55.6  55.8  56.0  55.9  56.0   

                       2013  2014  2015   2016  
Life expectancy                                 
Abkhazia                NaN   NaN   NaN    NaN  
Afghanistan            54.8  54.9  53.8  52.72  
Akrotiri and Dhekelia   NaN   NaN   NaN    NaN  
Albania                77.7  77.9  78.0  78.10  
Algeria                76.3  76.3  76.4  76.50  
...                     ...   ...   ...    ...  
Yugoslavia              NaN   NaN   NaN    NaN  
Zambia                 55.6  56.3  56.7  57.10  
Zimbabwe               55.7  57.0  59.3  61.69  
Åland                   NaN   NaN   NaN    NaN  
South Sudan            56.0  56.1  56.1  56.10  

[260 rows x 217 columns]

In [27]:
life.index

Index(['Abkhazia', 'Afghanistan', 'Akrotiri and Dhekelia', 'Albania',
       'Algeria', 'American Samoa', 'Andorra', 'Angola', 'Anguilla',
       'Antigua and Barbuda',
       ...
       'Vietnam', 'Virgin Islands (U.S.)', 'North Yemen (former)',
       'South Yemen (former)', 'Yemen', 'Yugoslavia', 'Zambia', 'Zimbabwe',
       'Åland', 'South Sudan'],
      dtype='object', name='Life expectancy', length=260)

In [28]:
life.index.name = 'country'

In [29]:
life = life.reset_index()

In [30]:
life = life.melt(id_vars='country', var_name='year', value_name='life_expectancy')

In [31]:
life

country  year  life_expectancy
0                   Abkhazia  1800              NaN
1                Afghanistan  1800            28.21
2      Akrotiri and Dhekelia  1800              NaN
3                    Albania  1800            35.40
4                    Algeria  1800            28.82
...                      ...   ...              ...
56415             Yugoslavia  2016              NaN
56416                 Zambia  2016            57.10
56417               Zimbabwe  2016            61.69
56418                  Åland  2016              NaN
56419            South Sudan  2016            56.10

[56420 rows x 3 columns]

## Step 9

In [32]:
pop = pd.read_excel('./data/gapminder_population.xlsx', index_col=0)

In [33]:
pop

1800.0     1810.0     1820.0  \
Total population                                                                
Abkhazia                                            NaN        NaN        NaN   
Afghanistan                                   3280000.0  3280000.0  3323519.0   
Akrotiri and Dhekelia                               NaN        NaN        NaN   
Albania                                        410445.0   423591.0   438671.0   
Algeria                                       2503218.0  2595056.0  2713079.0   
...                                                 ...        ...        ...   
Northern Marianas                                   NaN        NaN        NaN   
South Georgia and the South Sandwich Islands        NaN        NaN        NaN   
US Minor Outlying Islands                           NaN        NaN        NaN   
Virgin Islands                                      NaN        NaN        NaN   
West Bank                                           NaN        NaN        NaN   

                                                 1830.0     1840.0     1850.0  \
Total population                                                                
Abkhazia                                            NaN        NaN        NaN   
Afghanistan                                   3448982.0  3625022.0  3810047.0   
Akrotiri and Dhekelia                               NaN        NaN        NaN   
Albania                                        457234.0   478227.0   506889.0   
Algeria                                       2880355.0  3082721.0  3299305.0   
...                                                 ...        ...        ...   
Northern Marianas                                   NaN        NaN        NaN   
South Georgia and the South Sandwich Islands        NaN        NaN        NaN   
US Minor Outlying Islands                           NaN        NaN        NaN   
Virgin Islands                                      NaN        NaN        NaN   
West Bank                                           NaN        NaN        NaN   

                                                 1860.0     1870.0     1880.0  \
Total population                                                                
Abkhazia                                            NaN        NaN        NaN   
Afghanistan                                   3973968.0  4169690.0  4419695.0   
Akrotiri and Dhekelia                               NaN        NaN        NaN   
Albania                                        552800.0   610036.0   672544.0   
Algeria                                       3536468.0  3811028.0  4143163.0   
...                                                 ...        ...        ...   
Northern Marianas                                   NaN        NaN        NaN   
South Georgia and the South Sandwich Islands        NaN        NaN        NaN   
US Minor Outlying Islands                           NaN        NaN        NaN   
Virgin Islands                                      NaN        NaN        NaN   
West Bank                                           NaN        NaN        NaN   

                                                 1890.0  ...      2006.0  \
Total population                                         ...               
Abkhazia                                            NaN  ...         NaN   
Afghanistan                                   4710171.0  ...  25183615.0   
Akrotiri and Dhekelia                               NaN  ...     15700.0   
Albania                                        741688.0  ...   3050741.0   
Algeria                                       4525691.0  ...  33749328.0   
...                                                 ...  ...         ...   
Northern Marianas                                   NaN  ...         NaN   
South Georgia and the South Sandwich Islands        NaN  ...         NaN   
US Minor Outlying Islands                           NaN  ...         NaN   
Virgin Islands                                     

In [34]:
pop.shape

(275, 81)

In [35]:
pop.tail()

1800.0  1810.0  1820.0  1830.0  \
Total population                                                               
Northern Marianas                                NaN     NaN     NaN     NaN   
South Georgia and the South Sandwich Islands     NaN     NaN     NaN     NaN   
US Minor Outlying Islands                        NaN     NaN     NaN     NaN   
Virgin Islands                                   NaN     NaN     NaN     NaN   
West Bank                                        NaN     NaN     NaN     NaN   

                                              1840.0  1850.0  1860.0  1870.0  \
Total population                                                               
Northern Marianas                                NaN     NaN     NaN     NaN   
South Georgia and the South Sandwich Islands     NaN     NaN     NaN     NaN   
US Minor Outlying Islands                        NaN     NaN     NaN     NaN   
Virgin Islands                                   NaN     NaN     NaN     NaN   
West Bank                                        NaN     NaN     NaN     NaN   

                                              1880.0  1890.0  ...  2006.0  \
Total population                                              ...           
Northern Marianas                                NaN     NaN  ...     NaN   
South Georgia and the South Sandwich Islands     NaN     NaN  ...     NaN   
US Minor Outlying Islands                        NaN     NaN  ...     NaN   
Virgin Islands                                   NaN     NaN  ...     NaN   
West Bank                                        NaN     NaN  ...     NaN   

                                              2007.0  2008.0  2009.0  2010.0  \
Total population                                                               
Northern Marianas                                NaN     NaN     NaN     NaN   
South Georgia and the South Sandwich Islands     NaN     NaN     NaN     NaN   
US Minor Outlying Islands                        NaN     NaN     NaN     NaN   
Virgin Islands                                   NaN     NaN     NaN     NaN   
West Bank                                        NaN     NaN     NaN     NaN   

                                              2011.0  2012.0  2013.0  2014.0  \
Total population                                                               
Northern Marianas                                NaN     NaN     NaN     NaN   
South Georgia and the South Sandwich Islands     NaN     NaN     NaN     NaN   
US Minor Outlying Islands                        NaN     NaN     NaN     NaN   
Virgin Islands                                   NaN     NaN     NaN     NaN   
West Bank                                        NaN     NaN     NaN     NaN   

                                              2015.0  
Total population                                      
Northern Marianas                                NaN  
South Georgia and the South Sandwich Islands     NaN  
US Minor Outlying Islands                        NaN  
Virgin Islands                                   NaN  
West Bank                                        NaN  

[5 rows x 81 columns]

**Observation 5: The pop-Dateframe lists 15 more countries or regions then the orther two dataframes.**

**Observation 6: The pop-Dateframe includes only 81 columns compared to the 217 columns of the fert and life dataframes.**

In [36]:
pop.columns = pop.columns.astype(int)

## Tidy the pop dataframe

In [37]:
#Adjustment: Drop the last 15 rows, because these countries or regions don't show up in the other two data frames

In [38]:
pop = pop[:260]

In [39]:
pop.tail()

1800       1810       1820       1830       1840  \
Total population                                                          
Yugoslavia              NaN        NaN        NaN        NaN        NaN   
Zambia             747000.0   747000.0   747000.0   747000.0   747000.0   
Zimbabwe          1085814.0  1085814.0  1106634.0  1167259.0  1253642.0   
Åland                   NaN        NaN        NaN        NaN        NaN   
South Sudan       1027378.0  1041076.0  1063732.0  1103246.0  1153401.0   

                       1850       1860       1870       1880       1890  ...  \
Total population                                                         ...   
Yugoslavia              NaN        NaN        NaN        NaN        NaN  ...   
Zambia             747000.0   746052.0   745105.0   744157.0   743210.0  ...   
Zimbabwe          1346417.0  1444385.0  1549480.0  1661683.0  1781947.0  ...   
Åland                   NaN        NaN        NaN        NaN        NaN  ...   
South Sudan       1205835.0  1329906.0  1462765.0  1604755.0  1756589.0  ...   

                        2006        2007        2008        2009        2010  \
Total population                                                               
Yugoslavia        23887171.0  23946096.0  23988543.0         NaN         NaN   
Zambia            12381509.0  12738676.0  13114579.0  13507849.0  13917439.0   
Zimbabwe          13127942.0  13297798.0  13495462.0  13720997.0  13973897.0   
Åland                26923.0     26923.0     26923.0     26923.0     26923.0   
South Sudan        8445659.0   8815495.0   9208598.0   9623176.0  10056475.0   

                        2011        2012        2013        2014        2015  
Total population                                                              
Yugoslavia               NaN         NaN         NaN         NaN         NaN  
Zambia            14343526.0  14786581.0  15246086.0  15721343.0  16211767.0  
Zimbabwe          14255592.0  14565482.0  14898092.0  15245855.0  15602751.0  
Åland                    NaN         NaN         NaN         NaN         NaN  
South Sudan       10510122.0  10980623.0  11453810.0  11911184.0  12339812.0  

[5 rows x 81 columns]

## Convert the pop dataframe to the long format

In [40]:
pop.index.name = 'country'

In [41]:
pop = pop.reset_index()

In [42]:
pop = pop.melt(id_vars='country', var_name='year', value_name='population')

# Step 9

In [43]:
fert_pop = fert.merge(pop)

In [44]:
df = fert_pop.merge(life)

In [45]:
df

country  year  fertility_rate  population  \
0                   Abkhazia  1800             NaN         NaN   
1                Afghanistan  1800            7.00   3280000.0   
2      Akrotiri and Dhekelia  1800             NaN         NaN   
3                    Albania  1800            4.60    410445.0   
4                    Algeria  1800            6.99   2503218.0   
...                      ...   ...             ...         ...   
20974             Yugoslavia  2015             NaN         NaN   
20975                 Zambia  2015            5.59  16211767.0   
20976               Zimbabwe  2015            3.35  15602751.0   
20977                  Åland  2015             NaN         NaN   
20978            South Sudan  2015             NaN  12339812.0   

       life_expectancy  
0                  NaN  
1                28.21  
2                  NaN  
3                35.40  
4                28.82  
...                ...  
20974              NaN  
20975            56.70  
20976            59.30  
20977              NaN  
20978            56.10  

[20979 rows x 5 columns]

Realisation: It seems that the whole process of cleaning up the data frames was not necessary. For example, the year 2016 no longer appears at all. It seems as if the merge function automatically removes all columns that do not appear in both datasets. This is also reflected in the length of the final merged dataset "df".   It has only 20,979 rows compared to the 56,420 rows of the melted dataset "life".

# Step 10

In [46]:
#Add the continent data
cont = pd.read_csv('./data/continents.csv', sep=";")
df = df.merge(cont)

In [47]:
df.rename(columns={"country": "Country name",
                   "population": "Population size", 
                   "continent": "Continent",
                   "life_expectancy": "Life expectancy",
                   "fertility_rate": "Fertility rate"
                  }, inplace=True)

In [48]:
df

Country name  year  Fertility rate  Population size  Life expectancy  \
0      Afghanistan  1800             7.0        3280000.0            28.21   
1      Afghanistan  1810             7.0        3280000.0            28.11   
2      Afghanistan  1820             7.0        3323519.0            28.01   
3      Afghanistan  1830             7.0        3448982.0            27.90   
4      Afghanistan  1840             7.0        3625022.0            27.80   
...            ...   ...             ...              ...              ...   
14251  South Sudan  2011             NaN       10510122.0            55.90   
14252  South Sudan  2012             NaN       10980623.0            56.00   
14253  South Sudan  2013             NaN       11453810.0            56.00   
14254  South Sudan  2014             NaN       11911184.0            56.10   
14255  South Sudan  2015             NaN       12339812.0            56.10   

      Continent  
0          Asia  
1          Asia  
2          Asia  
3          Asia  
4          Asia  
...         ...  
14251    Africa  
14252    Africa  
14253    Africa  
14254    Africa  
14255    Africa  

[14256 rows x 6 columns]

# Step 11 and 12

In [59]:
for i in range(1960, 1961):
    dims = (12, 8)
    fig, ax = plt.subplots(figsize=dims)
    df_subset = df.loc[df['year'] == i]
    
    sns.set_style("whitegrid")
    ax = sns.scatterplot(x='Life expectancy', y='Fertility rate',
                data=df_subset, size='Population size', hue='Continent')
    sns.move_legend(ax, "upper right", fontsize=13)
    
    props = dict(boxstyle="round", facecolor='blue', alpha=0.1)
    ax.text(0.05, 0.07, str(i), transform=ax.transAxes, fontsize=14,
        verticalalignment='top', bbox=props)
    plt.title("Correlation between fertility rate and life expectancy (1960 - 2015)", fontsize=20, fontweight="bold")
    plt.xlabel("Life expectancy", fontsize=15)
    plt.ylabel("Fertility rate", fontsize=15)

    plt.xticks(fontsize=13)
    plt.yticks(fontsize=13)
        
    plt.axis((20,90,0,9))
    
    path = "./data/plots/lifeexp" + str(i) + ".png"
    plt.savefig(path)
    plt.close()

## Step 13

In [61]:
images = []

for i in range(1960, 2016):
    path = "./data/plots/lifeexp" + str(i) + ".png"
    images.append(imageio.imread(path))

imageio.mimsave('op_old_legend_10fps.gif', images, fps=10)

## Extra: Working on the way population is depicted in the legend

In [51]:
df_copy = df.copy()

In [52]:
df_copy

Country name  year  Fertility rate  Population size  Life expectancy  \
0      Afghanistan  1800             7.0        3280000.0            28.21   
1      Afghanistan  1810             7.0        3280000.0            28.11   
2      Afghanistan  1820             7.0        3323519.0            28.01   
3      Afghanistan  1830             7.0        3448982.0            27.90   
4      Afghanistan  1840             7.0        3625022.0            27.80   
...            ...   ...             ...              ...              ...   
14251  South Sudan  2011             NaN       10510122.0            55.90   
14252  South Sudan  2012             NaN       10980623.0            56.00   
14253  South Sudan  2013             NaN       11453810.0            56.00   
14254  South Sudan  2014             NaN       11911184.0            56.10   
14255  South Sudan  2015             NaN       12339812.0            56.10   

      Continent  
0          Asia  
1          Asia  
2          Asia  
3          Asia  
4          Asia  
...         ...  
14251    Africa  
14252    Africa  
14253    Africa  
14254    Africa  
14255    Africa  

[14256 rows x 6 columns]

In [53]:
row = df_copy["Population size"].tolist()

In [54]:
new_row = list()

for i in row:
    if i <= 100_000_000:
        new_row.append(0.1)
    elif i <= 500_000_000:
        new_row.append(0.5)
    else:
        new_row.append(1.5)

In [55]:
df_copy["Population in Billion (upto)"] = new_row

In [56]:
df_copy["Population in Billion (upto)"].value_counts()

0.1    13784
0.5      355
1.5      117
Name: Population in Billion (upto), dtype: int64

In [57]:
for i in range(1960, 2016):
    dims = (12, 8)
    fig, ax = plt.subplots(figsize=dims)
    df_subset = df_copy.loc[df['year'] == i]
    
    sns.set_style("whitegrid")
    ax = sns.scatterplot(x='Life expectancy', y='Fertility rate',
                data=df_subset, size="Population in Billion (upto)", hue='Continent')
    sns.move_legend(ax, "upper right", fontsize=13)
    
    props = dict(boxstyle="round", facecolor='blue', alpha=0.1)
    ax.text(0.05, 0.07, str(i), transform=ax.transAxes, fontsize=14,
        verticalalignment='top', bbox=props)
    plt.title("Correlation between fertility rate and life expectancy (1960 - 2015)", fontsize=20, fontweight="bold")
    plt.xlabel("Life expectancy", fontsize=15)
    plt.ylabel("Fertility rate", fontsize=15)

    plt.xticks(fontsize=13)
    plt.yticks(fontsize=13)
        
    plt.axis((20,90,0,9))
    
    path = "./data/plots/new/lifeexp_new" + str(i) + ".png"
    plt.savefig(path)
    plt.close()

In [63]:
images = []

for i in range(1960, 2016):
    path = "./data/plots/new/lifeexp_new" + str(i) + ".png"
    images.append(imageio.imread(path))

imageio.mimsave('op_new_legend_10fps.gif', images, fps=10)